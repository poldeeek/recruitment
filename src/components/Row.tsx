import { useState } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Box from '@material-ui/core/Box';
import Collapse from '@material-ui/core/Collapse';
import IconButton from '@material-ui/core/IconButton';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import Typography from '@material-ui/core/Typography';
import KeyboardArrowDownIcon from '@material-ui/icons/KeyboardArrowDown';
import KeyboardArrowUpIcon from '@material-ui/icons/KeyboardArrowUp';
import { useMediaQuery } from 'react-responsive';

import { ICourse } from '../types';

// remove duplicated bottom border in each row
const useRowStyles = makeStyles({
  root: {
    '& > *': {
      borderBottom: 'unset',
    },
  },
});

type RowProps = {
  course: ICourse;
};

const Row: React.FC<RowProps> = ({ course }) => {
  const [open, setOpen] = useState(false);
  const classes = useRowStyles();

  const isDesktopOrLaptop = useMediaQuery({
    query: '(min-device-width: 992px)',
  });

  const boxMargin = isDesktopOrLaptop ? 8 : 1;

  return (
    <>
      <TableRow className={classes.root}>
        <TableCell>
          <IconButton
            aria-label="expand row"
            size="small"
            onClick={() => setOpen(!open)}>
            {open ? (
              <KeyboardArrowUpIcon fontSize="large" />
            ) : (
              <KeyboardArrowDownIcon fontSize="large" />
            )}
          </IconButton>
        </TableCell>
        <TableCell component="th" scope="row">
          <Typography variant="h5" component="h5">
            {course.course}
          </Typography>
        </TableCell>
        <TableCell align="right">
          <Typography variant="h5" component="h5">
            {course.openedLessonsCount}
          </Typography>
        </TableCell>
      </TableRow>
      <TableRow>
        <TableCell style={{ paddingBottom: 0, paddingTop: 0 }} colSpan={6}>
          <Collapse in={open} timeout="auto" unmountOnExit>
            <Box marginX={boxMargin} marginY={1}>
              <Typography variant="h5" gutterBottom component="h5">
                Projects
              </Typography>
              <Table size="small" aria-label="purchases">
                <TableHead>
                  <TableRow>
                    <TableCell>
                      <Typography variant="h6" gutterBottom component="h6">
                        Name
                      </Typography>
                    </TableCell>
                    <TableCell align="right">
                      <Typography variant="h6" gutterBottom component="h6">
                        Completed Lessons
                      </Typography>
                    </TableCell>
                  </TableRow>
                </TableHead>
                <TableBody>
                  {course.projects.map((project) => (
                    <TableRow key={project.project}>
                      <TableCell component="th" scope="row">
                        <Typography variant="body1" gutterBottom>
                          {project.project}
                        </Typography>
                      </TableCell>
                      <TableCell align="right">
                        <Typography variant="body1" gutterBottom>
                          {project.completedLessonsCount}
                        </Typography>
                      </TableCell>
                    </TableRow>
                  ))}
                </TableBody>
              </Table>
            </Box>
          </Collapse>
        </TableCell>
      </TableRow>
    </>
  );
};

export default Row;
