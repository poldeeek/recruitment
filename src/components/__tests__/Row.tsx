import '@testing-library/jest-dom/extend-expect';
import Row from '../Row';
import { render, fireEvent } from '@testing-library/react';
import React from 'react';
import { act } from 'react-dom/test-utils';

const props = {
  course: 'Course1',
  openedLessonsCount: 3,
  projects: [
    {
      project: 'Poroject1',
      completedLessonsCount: 3,
    },
  ],
};

/*
    Wrapper for avoid that error in tests:
    Warning: validateDOMNesting(...): <tr> cannot appear as a child of <div>.
*/
const Wrapper: React.FC = ({ children }) => {
  return (
    <table>
      <tbody>{children}</tbody>
    </table>
  );
};

test('Basic render', () => {
  const component = render(
    <Wrapper>
      <Row course={props} />
    </Wrapper>
  );
  expect(component.baseElement).toMatchSnapshot();
});

test('Show projects', () => {
  const component = render(
    <Wrapper>
      <Row course={props} />
    </Wrapper>
  );
  const clickHanlder = component.getByRole('button');
  act(() => {
    fireEvent.click(clickHanlder);
  });
  expect(component.baseElement).toMatchSnapshot();
});
