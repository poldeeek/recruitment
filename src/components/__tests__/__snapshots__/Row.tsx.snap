// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`990px Basic render 1`] = `
<body>
  <div>
    <table>
      <tbody>
        <tr
          class="MuiTableRow-root makeStyles-root-1"
        >
          <td
            class="MuiTableCell-root"
          >
            <button
              aria-label="expand row"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
              role="button"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </td>
          <th
            class="MuiTableCell-root"
            role="cell"
            scope="row"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5"
            >
              Course1
            </h5>
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-alignRight"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5"
            >
              3
            </h5>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root"
            colspan="6"
            style="padding-bottom: 0px; padding-top: 0px;"
          />
        </tr>
      </tbody>
    </table>
  </div>
</body>
`;

exports[`1200px Basic render 1`] = `
<body>
  <div>
    <table>
      <tbody>
        <tr
          class="MuiTableRow-root makeStyles-root-2"
        >
          <td
            class="MuiTableCell-root"
          >
            <button
              aria-label="expand row"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
              role="button"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </td>
          <th
            class="MuiTableCell-root"
            role="cell"
            scope="row"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5"
            >
              Course1
            </h5>
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-alignRight"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5"
            >
              3
            </h5>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root"
            colspan="6"
            style="padding-bottom: 0px; padding-top: 0px;"
          />
        </tr>
      </tbody>
    </table>
  </div>
</body>
`;

exports[`Basic render 1`] = `
<body>
  <div>
    <table>
      <tbody>
        <tr
          class="MuiTableRow-root makeStyles-root-1"
        >
          <td
            class="MuiTableCell-root"
          >
            <button
              aria-label="expand row"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
              role="button"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </td>
          <th
            class="MuiTableCell-root"
            role="cell"
            scope="row"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5"
            >
              Course1
            </h5>
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-alignRight"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5"
            >
              3
            </h5>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root"
            colspan="6"
            style="padding-bottom: 0px; padding-top: 0px;"
          />
        </tr>
      </tbody>
    </table>
  </div>
</body>
`;

exports[`Show projects 1`] = `
<body>
  <div>
    <table>
      <tbody>
        <tr
          class="MuiTableRow-root makeStyles-root-3"
        >
          <td
            class="MuiTableCell-root"
          >
            <button
              aria-label="expand row"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
              role="button"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7.41 15.41L12 10.83l4.59 4.58L18 14l-6-6-6 6z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </td>
          <th
            class="MuiTableCell-root"
            role="cell"
            scope="row"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5"
            >
              Course1
            </h5>
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-alignRight"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5"
            >
              3
            </h5>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root"
            colspan="6"
            style="padding-bottom: 0px; padding-top: 0px;"
          >
            <div
              class="MuiCollapse-container"
              style="min-height: 0px; height: 0px; transition-duration: 0ms;"
            >
              <div
                class="MuiCollapse-wrapper"
              >
                <div
                  class="MuiCollapse-wrapperInner"
                >
                  <div
                    class="MuiBox-root MuiBox-root-4"
                  >
                    <h5
                      class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
                    >
                      Projects
                    </h5>
                    <table
                      aria-label="purchases"
                      class="MuiTable-root"
                    >
                      <thead
                        class="MuiTableHead-root"
                      >
                        <tr
                          class="MuiTableRow-root MuiTableRow-head"
                        >
                          <th
                            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
                            scope="col"
                          >
                            <h6
                              class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom"
                            >
                              Name
                            </h6>
                          </th>
                          <th
                            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall"
                            scope="col"
                          >
                            <h6
                              class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom"
                            >
                              Completed Lessons
                            </h6>
                          </th>
                        </tr>
                      </thead>
                      <tbody
                        class="MuiTableBody-root"
                      >
                        <tr
                          class="MuiTableRow-root"
                        >
                          <th
                            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                            role="cell"
                            scope="row"
                          >
                            <p
                              class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
                            >
                              Poroject1
                            </p>
                          </th>
                          <td
                            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall"
                          >
                            <p
                              class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
                            >
                              3
                            </p>
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
              </div>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</body>
`;
